bee-node/src/node.rs

```diff
@@ -9,49 +9,183 @@
+
+        info!("Reading snapshot file...");
+        let snapshot_state = match block_on(LocalSnapshot::from_file(self.config.snapshot.local().file_path())) {
+            Ok(local_snapshot) => {
+                info!(
+                    "Read snapshot file from {} with index {}, {} solid entry points, {} seen milestones and {} balances.",
+                    Utc.timestamp(local_snapshot.metadata().timestamp() as i64, 0)
+                        .to_rfc2822(),
+                    local_snapshot.metadata().index(),
+                    local_snapshot.metadata().solid_entry_points().len(),
+                    local_snapshot.metadata().seen_milestones().len(),
+                    local_snapshot.state().balances().len()
+                );
+
+                tangle::tangle().update_solid_milestone_index(local_snapshot.metadata().index().into());
+
+                // TODO get from database
+                tangle::tangle().update_last_milestone_index(local_snapshot.metadata().index().into());
+
+                tangle::tangle().update_snapshot_milestone_index(local_snapshot.metadata().index().into());
+
+                // TODO index 0 ?
+                tangle::tangle().add_solid_entry_point(Hash::zeros(), MilestoneIndex(0));
+                for (hash, index) in local_snapshot.metadata().solid_entry_points() {
+                    tangle::tangle().add_solid_entry_point(*hash, MilestoneIndex(*index));
+                }
+
+                for _seen_milestone in local_snapshot.metadata().seen_milestones() {
+                    // TODO request ?
+                }
+
+                local_snapshot.into_state()
+            }
+            Err(e) => {
+                error!(
+                    "Failed to read snapshot file \"{}\": {:?}.",
+                    self.config.snapshot.local().file_path(),
+                    e
+                );
+                return Err(Error::SnapshotReadError(e));
+            }
+        };
+
+        block_on(Protocol::init(self.config.protocol.clone(), network.clone()));
+
+        // TODO config
+        let (ledger_worker_tx, ledger_worker_rx) = mpsc::channel(1000);
+        let (ledger_worker_shutdown_tx, ledger_worker_shutdown_rx) = oneshot::channel();
+
+        spawn(LedgerWorker::new(snapshot_state.into_balances()).run(ledger_worker_rx, ledger_worker_shutdown_rx));
+
+        info!("Initialized.");
+
+        Ok(Node {
+            config: self.config,
+            network,
+            events: events.fuse(),
+            shutdown,
+            ledger: (ledger_worker_tx, ledger_worker_shutdown_tx),
+            peers: HashMap::new(),
+        })
+    }
+}
 
+/// The main node type.
 pub struct Node {
     config: NodeConfig,
+    // TODO those 2 fields are related; consider bundling them
     network: Network,
-    events: EventSubscriber,
+    events: Fuse<EventSubscriber>,
     shutdown: Shutdown,
```

Comment:
Why do we fuse events and not shutdown here ?  For consistency with the other workers I'd fuse both in the run_loop...

Reply:
we fuse `events` because we only ever want it fused, hence we shouldn't be less restrictive on the field type. You are right, I should fuse the shutdown as well.

Reply:
actually....`Shutdown` is the struct from `bee-common`, and not a fusable stream or future. What you mean is the channel that listens for the Ctrl-C event. I would require an additional field in the `struct`. Do you prefer that?

Reply:
Ah no sorry I confused myself, just forget. I still would have fused the events where we need it but it's fine :)
You can close.